File: tools/server_observer/server_observer.py
"""
ServerObserver tool for lightweight recording of server responses across multiple games.
"""
from __future__ import annotations

import gc
from copy import deepcopy
from pathlib import Path
from queue import Empty
from queue import Queue
from threading import Event
from threading import Lock
from threading import Thread
from threading import current_thread
from time import time
from typing import Dict
from typing import Iterable
from typing import List
from typing import Optional
from typing import Set

from conflict_interface.data_types.hub_types.hub_game import HubGameProperties
from conflict_interface.data_types.hub_types.hub_game_state_enum import HubGameState
from conflict_interface.interface.hub_interface import HubInterface
from tools.server_observer.account import Account
from tools.server_observer.account_pool import AccountPool
from tools.server_observer.observation_session import ObservationSession
from tools.server_observer.recorder_logger import get_logger
from tools.server_observer.recording_registry import RecordingRegistry
from tools.server_observer.static_map_cache import StaticMapCache

logger = get_logger()
THREAD_JOIN_TIMEOUT = 1.0
MAX_UPDATE_RETRIES = 3


class ServerObserver:
    """
    Manage observation of multiple games concurrently, recording server responses.

    Uses an AccountPool to select guest accounts, scans hub listings similar to
    MultiRecorder, and schedules ObservationSessions on a thread pool. Intended
    for lightweight response logging without executing complex Recorder actions.
    """

    def __init__(self, config: dict, account_pool: AccountPool):
        self.config = config
        self.account_pool = account_pool

        self.scenario_ids: List[int] = config.get("scenario_ids", [])
        self.max_parallel_recordings: int = int(config.get("max_parallel_recordings", 1))
        self.max_parallel_updates: int = int(config.get("max_parallel_updates", 1))
        self.max_parallel_first_updates: int = int(config.get("max_parallel_first_updates", 1))
        self.scan_interval: float = float(config.get("scan_interval", 30))
        self.max_guest_per_account: Optional[int] = config.get("max_guest_games_per_account")
        self.update_interval: float = float(config.get("update_interval", 60.0))
        self.output_dir = Path(config.get("output_dir", "./recordings"))
        # Optional separate directory for metadata
        self.output_metadata_dir = config.get("output_metadata_dir")
        if self.output_metadata_dir is not None:
            self.output_metadata_dir = Path(self.output_metadata_dir)
        self.enabled_scanning = config.get("enabled_scanning", True)
        
        # Long-term storage configuration
        self.long_term_storage_path = config.get("long_term_storage_path")
        if self.long_term_storage_path is not None:
            self.long_term_storage_path = Path(self.long_term_storage_path)
        self.file_size_threshold = config.get("file_size_threshold")

        self.observer_sessions: Dict[int, ObservationSession] = {}

        # Use metadata directory for registry if configured
        registry_default_dir = self.output_metadata_dir if self.output_metadata_dir else self.output_dir
        registry_path = config.get(
            "registry_path",
            registry_default_dir / "server_observer_registry.json",
        )
        self.registry = RecordingRegistry(Path(registry_path))
        self._listing_interface: Optional[HubInterface] = None
        self._listing_account: Optional[Account] = None
        self._active_threads: Set[Thread] = set()
        self._first_update_sessions: Set[int] = set()
        self._threads_lock = Lock()
        self._stop_event = Event()
        self._update_queue: Queue[ObservationSession] = Queue()
        self._scan_thread: Optional[Thread] = None
        self._map_cache = StaticMapCache(self.output_dir / "static_maps")
        self._known_games: Set[int] = set()
        self._refresh_known_games_from_registry()

    def _get_listing_interface(self) -> Optional[HubInterface]:
        if self._listing_interface:
            return self._listing_interface

        try:
            if self.account_pool is None:
                raise Exception("No account pool provided")
            account = self.account_pool.get_any_account() or self.account_pool.next_free_account()
            if not account:
                logger.error("No accounts available for listing games")
                return None
            self._listing_account = account
            self._listing_interface = account.get_interface()
            return self._listing_interface
        except Exception as exc:
            logger.error(f"Failed to prepare listing interface: {exc}")
            return None

    def _select_games(self, interface: HubInterface) -> Iterable[tuple[int, HubGameProperties]]:
        logger.info("Scanning for games")
        games = interface.get_global_games()
        seen_games: set[int] = set()

        for scenario_id in self.scenario_ids:
            new_candidates = [
                game for game in games
                if game.scenario_id == scenario_id and game.game_id not in self._known_games
            ]
            joinable = [g for g in new_candidates if g.open_slots >= 1]

            logger.info(
                f"Scenario {scenario_id}: {len(new_candidates)} new games, "
                f"{len(joinable)} potentially joinable before sampling"
            )

            for game in joinable:
                if game.game_id in seen_games:
                    continue
                seen_games.add(game.game_id)
                yield scenario_id, game

    def _refresh_known_games_from_registry(self):
        self._known_games = {
            int(gid)
            for bucket in self.registry.state.values()
            for gid in bucket.keys()
        }

    def _pick_account(self) -> Optional[Account]:
        if not self.account_pool:
            return None
        tried = 0
        total = len(self.account_pool.accounts)
        while tried < total:
            account = self.account_pool.next_guest_account(self.max_guest_per_account)
            if account is None:
                return None
            if account != self._listing_account:
                return account
            else:
                self.account_pool.guest_account_pointer += 1
            tried += 1
        logger.error("No non-listing account available to start new observation")
        return None

    def _start_observation_session(self, game_id: int, scenario_id: int):
        account = self._pick_account()
        if self.account_pool and not account:
            logger.error("No free account available to start new observation")
            return
        logger.info(f"Starting observation session for game {game_id} with scenario {scenario_id}")

        # Determine metadata path for this game
        metadata_path = None
        if self.output_metadata_dir is not None:
            metadata_path = self.output_metadata_dir / f"game_{game_id}"

        observer = ObservationSession(
            game_id,
            account,
            self._map_cache,
            self.output_dir/f"game_{game_id}",
            metadata_path=metadata_path,
            long_term_storage_path=self.long_term_storage_path,
            file_size_threshold=self.file_size_threshold
        )
        self.registry.mark_recording(game_id, scenario_id, None)
        self._known_games.add(game_id)
        self.account_pool.increment_guest_join(account)
        self.observer_sessions[game_id] = observer
        observer.next_update_at = time()
        with self._threads_lock:
            self._first_update_sessions.add(game_id)
        self._update_queue.put(observer)

    def _resume_active(self):
        for game_id, meta in self.registry.active().items():
            scenario_id = meta.get("scenario_id")
            if scenario_id is None:
                logger.warning(f"Skipping resume for game {game_id} without scenario metadata")
                continue
            if game_id in self.observer_sessions:
                continue
            logger.info(f"Resuming observation for game {game_id}")
            if len(self.observer_sessions) >= self.max_parallel_recordings:
                logger.warning(f"Skipping resume for game {game_id} due to max parallel limit")
                continue
            self._start_observation_session(game_id, scenario_id)

    def _run_single_update(self, session: ObservationSession):
        game_id = session.game_id
        attempt = 1
        while True:
            try:
                with session.create_worker() as worker:
                    keep_running = worker.run()
                    session.update_package(deepcopy(worker.package))

                    if keep_running:
                        session.next_update_at = time() + self.update_interval
                        self._update_queue.put(session)
                    else:
                        self.registry.mark_completed(game_id)
                        self.account_pool.decrement_guest_join(session.account)
                        self.observer_sessions.pop(game_id, None)
                        self._known_games.add(game_id)
                with self._threads_lock:
                    self._first_update_sessions.discard(game_id)
                break  # Success, exit the retry loop
            except Exception as e:
                if attempt < MAX_UPDATE_RETRIES:
                    logger.exception(f"Observation for game {game_id} failed, retrying attempt {attempt}/{MAX_UPDATE_RETRIES}...")
                    attempt += 1
                    continue

                logger.exception(f"Observation for game {game_id} failed after {MAX_UPDATE_RETRIES} retries, marking as failed.")
                self.registry.mark_failed(game_id, str(e))
                self.account_pool.decrement_guest_join(session.account)
                self.observer_sessions.pop(game_id, None)
                self._known_games.add(game_id)
                with self._threads_lock:
                    self._first_update_sessions.discard(game_id)
                break  # Exit the retry loop
            finally:
                with self._threads_lock:
                    self._active_threads.discard(current_thread())
                gc.collect()

    def _start_due_updates(self):
        now = time()
        deferred: List[ObservationSession] = []
        while True:
            with self._threads_lock:
                if len(self._active_threads) >= self.max_parallel_updates:
                    break
            try:
                observer = self._update_queue.get_nowait()
            except Empty:
                break
            if not observer.needs_update(now):
                deferred.append(observer)
                continue

            # Check if this is a first-update and if we've hit the first-update limit
            with self._threads_lock:
                is_first_update = observer.game_id in self._first_update_sessions
                # Create a snapshot to avoid "set changed size during iteration" error
                first_update_snapshot = set(self._first_update_sessions)
                active_thread_names = {t.name for t in self._active_threads}

            if is_first_update:
                active_first_updates = sum(
                    1 for gid in first_update_snapshot
                    if self.observer_sessions.get(gid) and f"observer-{gid}" in active_thread_names
                )
                if active_first_updates >= self.max_parallel_first_updates:
                    deferred.append(observer)
                    continue

            thread = Thread(target=self._run_single_update, args=(observer,), name=f"observer-{observer.game_id}", daemon=True)
            logger.info(f"Starting ObserverWorker thread for game {observer.game_id}")
            with self._threads_lock:
                self._active_threads.add(thread)
            thread.start()
        for observer in deferred:
            self._update_queue.put(observer)
        if deferred:
            wait_times = [max(0.0, obs.next_update_at - now) for obs in deferred]
            wait_time = min(wait_times) if wait_times else 0.0
            if wait_time:
                self._stop_event.wait(min(wait_time, self.scan_interval))

    def _clean_finished_threads(self):
        with self._threads_lock:
            finished = [t for t in self._active_threads if not t.is_alive()]
            for thread in finished:
                thread.join(timeout=THREAD_JOIN_TIMEOUT)
                self._active_threads.discard(thread)

    def _scan_loop(self):
        interface = self._get_listing_interface()
        while not self._stop_event.is_set():
            if interface and self.enabled_scanning:
                for scenario_id, game in self._select_games(interface):
                    if game.game_id in self.observer_sessions:
                        continue
                    if len(self.observer_sessions) >= self.max_parallel_recordings:
                        continue
                    self._start_observation_session(game.game_id, scenario_id)
            self._stop_event.wait(self.scan_interval)

    def run(self) -> bool:
        self._stop_event.clear()
        self._resume_active()
        self._scan_thread = Thread(target=self._scan_loop, name="observer-scan", daemon=True)
        self._scan_thread.start()
        try:
            while not self._stop_event.is_set():
                self._clean_finished_threads()
                self._start_due_updates()
            return True
        finally:
            self._stop_event.set()
            if self._scan_thread:
                self._scan_thread.join()
            with self._threads_lock:
                threads = list(self._active_threads)
            for thread in threads:
                thread.join()
            self._clean_finished_threads()

File: tools/server_observer/src/server_observer.rs
use crate::account_pool::AccountPool;
use crate::db::{DbClient, DbConfig};
use crate::game_finder::{GameFinder, GameFinderConfig};
use crate::metrics::{
    record_game_completed, record_game_failed, record_game_started, record_missed_interval,
    record_scheduled_update_latency, record_game_update_completed, record_game_update_started,
    set_active_games,
};
use crate::observation_session::{ObservationError, ObservationResult, ObservationSession};
use crate::recording_registry::RecordingRegistry;
use crate::redis_publisher::{RedisConfig, RedisPublisher};
use crate::s3_client::{S3Client, S3Config};
use crate::scheduler::Scheduler;
use crate::static_map_cache::{StaticMapCache, StaticMapError};
use config::Config as AppConfig;
use std::collections::HashMap;
use std::sync::atomic::{AtomicBool, AtomicI32, Ordering};
use std::sync::{Arc, Mutex};
use std::time::SystemTime;
use thiserror::Error;

const MAX_UPDATE_RETRIES: i32 = 3;

#[derive(Debug, Error)]
pub enum ServerObserverError {
    #[error("map cache error: {0}")]
    StaticMap(#[from] StaticMapError),
    #[error("database error: {0}")]
    Database(#[from] crate::db::DbClientError),
    #[error("s3 error: {0}")]
    S3(#[from] crate::s3_client::S3Error),
    #[error("redis error: {0}")]
    Redis(#[from] crate::redis_publisher::RedisPublisherError),
}

pub struct ServerObserver {
    account_pool: Arc<tokio::sync::Mutex<AccountPool>>,
    scheduler: Arc<Scheduler>,
    max_parallel_recordings: AtomicI32,
    update_interval: Mutex<f64>,
    output_dir: String,
    output_metadata_dir: String,
    long_term_storage_path: String,
    file_size_threshold: i64,
    observer_sessions: Mutex<HashMap<i32, Arc<tokio::sync::Mutex<ObservationSession>>>>,
    registry: RecordingRegistry,
    game_finder: Mutex<Option<GameFinder>>,
    stop_flag: AtomicBool,
    map_cache: StaticMapCache,
    redis_publisher: Option<RedisPublisher>,
}

impl ServerObserver {
    pub async fn new(
        settings: &AppConfig,
        account_pool: Arc<tokio::sync::Mutex<AccountPool>>,
    ) -> Result<Arc<Self>, ServerObserverError> {
        let max_parallel_recordings = settings
            .get::<i64>("max_parallel_recordings")
            .unwrap_or(1) as i32;
        let max_parallel_updates = settings
            .get::<i64>("max_parallel_updates")
            .unwrap_or(1) as i32;
        let max_parallel_first_updates = settings
            .get::<i64>("max_parallel_first_updates")
            .unwrap_or(1) as i32;
        let update_interval = settings
            .get::<f64>("update_interval")
            .unwrap_or(60.0);

        let output_dir = settings
            .get::<String>("output_dir")
            .unwrap_or_else(|_| "./recordings".to_string());
        let output_metadata_dir = settings
            .get::<String>("output_metadata_dir")
            .unwrap_or_default();
        let long_term_storage_path = settings
            .get::<String>("long_term_storage_path")
            .unwrap_or_default();
        let file_size_threshold = settings
            .get::<i64>("file_size_threshold")
            .unwrap_or(0);

        let registry_default_dir = if output_metadata_dir.is_empty() {
            output_dir.clone()
        } else {
            output_metadata_dir.clone()
        };
        let registry_path = settings
            .get::<String>("registry_path")
            .unwrap_or_else(|_| format!("{registry_default_dir}/server_observer_registry.json"));
        let registry = RecordingRegistry::new(registry_path);

        let db_client = match settings.get::<DbConfig>("database") {
            Ok(cfg) => match DbClient::new(cfg).await {
                Ok(client) if client.is_connected().await => Some(client),
                Ok(_) => {
                    tracing::warn!(
                        "database configured but connection test failed; disabling DB features"
                    );
                    None
                }
                Err(err) => {
                    tracing::warn!(
                        ?err,
                        "failed to initialize database client; disabling DB features"
                    );
                    None
                }
            },
            Err(_) => None,
        };

        let static_maps_dir = settings
            .get::<String>("storage.static_maps_dir")
            .unwrap_or_else(|_| format!("{output_dir}/static_maps"));

        let s3_client = match settings.get::<S3Config>("storage.s3") {
            Ok(parsed) => Some(S3Client::new(parsed).await?),
            Err(_) => None,
        };

        let s3_enabled = s3_client.is_some();

        let map_cache = StaticMapCache::new(static_maps_dir, s3_client, db_client.clone()).await?;

        let redis_publisher = match settings.get::<RedisConfig>("redis") {
            Ok(cfg) => match RedisPublisher::new(cfg) {
                Ok(redis) => Some(redis),
                Err(err) => {
                    tracing::warn!(
                        ?err,
                        "failed to initialize redis publisher; disabling redis"
                    );
                    None
                }
            },
            Err(_) => None,
        };

        let db_enabled = db_client.is_some();
        let redis_enabled = redis_publisher.is_some();

        let scheduler = Arc::new(Scheduler::new(
            max_parallel_updates.max(1),
            max_parallel_first_updates.max(1),
            update_interval,
        ));
        scheduler.set_update_interval(update_interval);
        scheduler.set_max_parallel_updates(max_parallel_updates.max(1));
        scheduler.set_max_parallel_first_updates(max_parallel_first_updates.max(1));

        let enabled_scanning = settings
            .get::<bool>("game_finder.enabled")
            .unwrap_or(false);

        let scan_interval = settings
            .get::<f64>("game_finder.scan_interval_seconds")
            .unwrap_or(300.0);

        let max_parallel_recordings_cfg = settings
            .get::<i64>("game_finder.max_games_per_scan")
            .unwrap_or(max_parallel_recordings as i64) as i32;

        let scenario_ids = settings
            .get::<Vec<i32>>("game_finder.scenario_ids")
            .unwrap_or_else(|_| Vec::new());

        let max_guest_games_per_account = settings
            .get::<i64>("game_finder.max_guest_games_per_account")
            .unwrap_or(-1) as i32;

        let game_finder_cfg = GameFinderConfig {
            scenario_ids,
            scan_interval,
            enabled_scanning,
            max_parallel_recordings: max_parallel_recordings_cfg,
            max_guest_games_per_account,
        };
        let mut game_finder =
            GameFinder::new(game_finder_cfg.clone(), Arc::clone(&account_pool), registry.clone());
        game_finder.set_scan_interval(game_finder_cfg.scan_interval);
        game_finder.set_scenario_ids(game_finder_cfg.scenario_ids.clone());
        game_finder.set_max_parallel_recordings(game_finder_cfg.max_parallel_recordings);
        game_finder.set_max_guest_games_per_account(game_finder_cfg.max_guest_games_per_account);

        let observer = Arc::new(Self {
            account_pool,
            scheduler,
            max_parallel_recordings: AtomicI32::new(max_parallel_recordings.max(1)),
            update_interval: Mutex::new(update_interval),
            output_dir,
            output_metadata_dir,
            long_term_storage_path,
            file_size_threshold,
            observer_sessions: Mutex::new(HashMap::new()),
            registry,
            game_finder: Mutex::new(Some(game_finder)),
            stop_flag: AtomicBool::new(false),
            map_cache,
            redis_publisher,
        });

        tracing::info!(
            max_parallel_recordings,
            max_parallel_updates,
            max_parallel_first_updates,
            update_interval,
            db_enabled,
            s3_enabled,
            redis_enabled,
            output_dir = %observer.output_dir,
            output_metadata_dir = %observer.output_metadata_dir,
            long_term_storage_path = %observer.long_term_storage_path,
            file_size_threshold = observer.file_size_threshold,
            "initialized ServerObserver configuration"
        );

        observer.install_game_finder_callbacks();
        observer.install_account_pool_callbacks().await;
        Ok(observer)
    }

    pub async fn run(self: Arc<Self>) -> bool {
        self.stop_flag.store(false, Ordering::SeqCst);

        let current_max_parallel = self.max_parallel_recordings.load(Ordering::SeqCst);
        let current_update_interval = *self
            .update_interval
            .lock()
            .expect("update interval mutex poisoned");
        tracing::info!(
            max_parallel_recordings = current_max_parallel,
            update_interval = current_update_interval,
            "server observer run loop started"
        );

        self.resume_active().await;
        {
            let mut guard = self.game_finder.lock().expect("game finder mutex poisoned");
            if let Some(game_finder) = guard.as_mut() {
                game_finder.refresh_known_games_from_registry();
                let enabled = game_finder.is_scanning_enabled();
                game_finder.set_enabled_scanning(enabled);
            }
        }

        while !self.stop_flag.load(Ordering::SeqCst) {
            let pending = self.scheduler.get_pending_new_sessions();
            for (game_id, scenario_id) in pending {
                let can_start = {
                    let sessions = self
                        .observer_sessions
                        .lock()
                        .expect("observer sessions mutex poisoned");
                    sessions.len() < self.max_parallel_recordings.load(Ordering::SeqCst) as usize
                };
                if can_start {
                    self.start_observation_session(game_id, scenario_id).await;
                }
            }

            self.start_due_updates().await;
        }

        tracing::info!("server observer run loop stopping");

        true
    }

    pub async fn stop(&self) {
        self.stop_flag.store(true, Ordering::SeqCst);
        tracing::info!("stop requested; shutting down scheduler and game finder");
        self.scheduler.stop();

        let mut game_finder = {
            let mut guard = self.game_finder.lock().expect("game finder mutex poisoned");
            guard.take()
        };
        if let Some(ref mut finder) = game_finder {
            finder.stop_scanning().await;
        }

        self.observer_sessions
            .lock()
            .expect("observer sessions mutex poisoned")
            .clear();
    }

    async fn start_observation_session(&self, game_id: i32, scenario_id: i32) {
        let account = {
            let mut pool = self.account_pool.lock().await;
            pool.next_guest_account_owned(-1)
        };
        let Some(account) = account else {
            tracing::warn!("no free account available for new observation");
            return;
        };

        tracing::info!(
            game_id,
            scenario_id,
            account = %account.username,
            "starting new observation session"
        );

        let metadata_path = if self.output_metadata_dir.is_empty() {
            String::new()
        } else {
            format!("{}/game_{}", self.output_metadata_dir, game_id)
        };

        let mut session = ObservationSession::new(
            game_id,
            account.clone(),
            Some(self.map_cache.clone()),
            format!("{}/game_{}", self.output_dir, game_id),
            metadata_path,
            self.long_term_storage_path.clone(),
            self.file_size_threshold,
            self.redis_publisher.clone(),
        );

        self.registry.mark_recording(game_id, scenario_id, Some(""));
        self.scheduler.initialize_session_schedule(&mut session);
        self.scheduler.mark_first_update(game_id);
        self.scheduler.schedule_update(&session);

        {
            let mut sessions = self
                .observer_sessions
                .lock()
                .expect("observer sessions mutex poisoned");
            sessions.insert(game_id, Arc::new(tokio::sync::Mutex::new(session)));
        }

        {
            let mut pool = self.account_pool.lock().await;
            pool.increment_guest_join(&account.username);
        }

        {
            let mut game_finder = self.game_finder.lock().expect("game finder mutex poisoned");
            if let Some(finder) = game_finder.as_mut() {
                finder.mark_game_known(game_id);
            }
        }

        // Record game started metric and update active games gauge
        record_game_started(scenario_id);
        self.update_active_games_metrics().await;
    }

    async fn resume_active(&self) {
        let active = self.registry.active();
        tracing::info!(active_count = active.len(), "resuming active recordings from registry");
        let mut resumed = 0usize;
        for (game_id, meta) in active {
            let scenario_id = meta
                .get("scenario_id")
                .and_then(serde_json::Value::as_i64)
                .unwrap_or(-1) as i32;
            if scenario_id < 0 {
                tracing::warn!(game_id, "skipping registry entry with missing scenario_id");
                continue;
            }

            let already_exists = self
                .observer_sessions
                .lock()
                .expect("observer sessions mutex poisoned")
                .contains_key(&game_id);
            if already_exists {
                tracing::debug!(game_id, "skipping resume for game already in observer_sessions");
                continue;
            }

            if self
                .observer_sessions
                .lock()
                .expect("observer sessions mutex poisoned")
                .len()
                >= self.max_parallel_recordings.load(Ordering::SeqCst) as usize
            {
                tracing::warn!(
                    "reached max_parallel_recordings while resuming; remaining sessions will be picked up later"
                );
                break;
            }
            self.start_observation_session(game_id, scenario_id).await;
            resumed += 1;
        }
        tracing::info!(resumed, "finished resuming active recordings");
    }

    async fn start_due_updates(self: &Arc<Self>) {
        let due_sessions = self.scheduler.get_due_updates();
        if !due_sessions.is_empty() {
            tracing::debug!(due_count = due_sessions.len(), "starting due updates");
        }
        for game_id in due_sessions {
            self.scheduler.increment_active_coroutines();
            let observer = Arc::clone(self);
            tokio::spawn(async move {
                observer.run_single_update_async(game_id).await;
            });
        }
        self.scheduler.process_due_updates().await;
    }

    async fn run_single_update_async(self: Arc<Self>, game_id: i32) {
        let session_arc = {
            self.observer_sessions
                .lock()
                .expect("observer sessions mutex poisoned")
                .get(&game_id)
                .cloned()
        };
        let Some(session_arc) = session_arc else {
            tracing::debug!(game_id, "due update skipped; session no longer exists");
            self.scheduler.decrement_active_coroutines();
            return;
        };

        // Record game update started and scheduled latency metrics
        let scheduled_latency_secs = {
            let session_opt = {
                self.observer_sessions
                    .lock()
                    .expect("observer sessions mutex poisoned")
                    .get(&game_id)
                    .cloned()
            };
            if let Some(session_arc) = session_opt {
                let session = session_arc.lock().await;
                let now = SystemTime::now();
                if now > session.next_update_at {
                    now.duration_since(session.next_update_at)
                        .map(|d| d.as_secs_f64())
                        .ok()
                } else {
                    None
                }
            } else {
                None
            }
        };

        record_game_update_started();
        if let Some(latency) = scheduled_latency_secs {
            record_scheduled_update_latency(latency);
            if latency > 10.0 {
                record_missed_interval();
            }
        }

        let result = {
            let mut session = session_arc.lock().await;
            session.run_update_async().await
        };

        if result.game_ended {
            self.handle_game_ended(game_id, &session_arc).await;
        } else if result.error_code == ObservationError::Success {
            self.handle_successful_update(&session_arc).await;
        } else {
            self.handle_failed_update(game_id, &session_arc, result).await;
        }

        self.scheduler.cleanup_first_update_tracking(game_id);
        record_game_update_completed();
        self.scheduler.decrement_active_coroutines();
    }

    async fn handle_game_ended(
        &self,
        game_id: i32,
        session_arc: &Arc<tokio::sync::Mutex<ObservationSession>>,
    ) {
        let scenario_id = self.registry.get_scenario_id(game_id);
        tracing::info!(game_id, ?scenario_id, "game ended, completing recording");
        self.registry.mark_completed(game_id);
        let username = session_arc.lock().await.account.username.clone();
        {
            let mut pool = self.account_pool.lock().await;
            pool.decrement_guest_join(&username);
        }
        self.observer_sessions
            .lock()
            .expect("observer sessions mutex poisoned")
            .remove(&game_id);

        if let Some(scenario_id) = scenario_id {
            record_game_completed(scenario_id);
        }

        self.update_active_games_metrics().await;
    }

    async fn handle_successful_update(&self, session_arc: &Arc<tokio::sync::Mutex<ObservationSession>>) {
        let mut session = session_arc.lock().await;
        let now = SystemTime::now();
        let latency_secs = now
            .duration_since(session.next_update_at)
            .map(|d| d.as_secs_f64())
            .unwrap_or(0.0);
        let update_interval = *self.update_interval.lock().expect("update interval mutex poisoned");
        let missed_update = latency_secs > update_interval;

        tracing::info!(
            game_id = session.game_id,
            latency_secs,
            missed_update,
            "successful update completed"
        );

        self.scheduler
            .schedule_next_update(&mut session, missed_update);
        session.reset_attempt();
    }

    async fn handle_failed_update(
        &self,
        game_id: i32,
        session_arc: &Arc<tokio::sync::Mutex<ObservationSession>>,
        result: ObservationResult,
    ) {
        let mut drop_session = false;
        let username: String;
        let mut needs_proxy_reset = false;

        {
            let mut session = session_arc.lock().await;
            username = session.account.username.clone();
            if session.get_attempt() >= MAX_UPDATE_RETRIES {
                drop_session = true;
            } else {
                session.increment_attempt();
                needs_proxy_reset = result.error_code == ObservationError::NetworkError;
                let immediate = self.should_retry_immediately(result.error_code);
                tracing::warn!(
                    game_id,
                    attempt = session.get_attempt(),
                    ?result.error_code,
                    immediate_retry = immediate,
                    "update failed; scheduling retry"
                );
                if immediate {
                    self.scheduler.schedule_immediate_update(&mut session);
                } else {
                    self.scheduler.schedule_next_update(&mut session, false);
                }
            }
        }

        if needs_proxy_reset {
            let mut pool = self.account_pool.lock().await;
            if !pool.reset_account_proxy(&username).await {
                tracing::warn!(account = username, "failed to reset proxy after network error");
            }
        }

        if drop_session {
            tracing::error!(
                game_id,
                account = username,
                error_message = %result.error_message,
                "dropping observation session after exceeding max retries"
            );
            self.registry.mark_failed(game_id, Some(&result.error_message));
            {
                let mut pool = self.account_pool.lock().await;
                pool.decrement_guest_join(&username);
            }
            self.observer_sessions
                .lock()
                .expect("observer sessions mutex poisoned")
                .remove(&game_id);
            let error_type = match result.error_code {
                ObservationError::AuthFailed => "auth_failed",
                ObservationError::ServerError => "server_error",
                ObservationError::NetworkError => "network_error",
                ObservationError::PackageCreationFailed => "package_creation_failed",
                ObservationError::UnknownError => "unknown_error",
                ObservationError::Success | ObservationError::GameEnded => "unknown_error",
            };
            record_game_failed(error_type);

            self.update_active_games_metrics().await;
        }
    }

    async fn update_active_games_metrics(&self) {
        let mut active_by_scenario = std::collections::HashMap::<i32, i64>::new();
        {
            let sessions = self
                .observer_sessions
                .lock()
                .expect("observer sessions mutex poisoned");
            for (&game_id, _) in sessions.iter() {
                if let Some(scenario_id) = self.registry.get_scenario_id(game_id) {
                    *active_by_scenario.entry(scenario_id).or_insert(0) += 1;
                }
            }
        }

        for (scenario_id, count) in active_by_scenario {
            set_active_games(scenario_id, count);
        }
    }

    fn should_retry_immediately(&self, error_code: ObservationError) -> bool {
        error_code == ObservationError::AuthFailed || error_code == ObservationError::ServerError
    }

    async fn reset_sessions_for_account(
        &self,
        username: &str,
        new_proxy: crate::account_pool::ProxyConfig,
    ) {
        let session_arcs = {
            self.observer_sessions
                .lock()
                .expect("observer sessions mutex poisoned")
                .values()
                .cloned()
                .collect::<Vec<_>>()
        };

        for session_arc in session_arcs {
            let mut session = session_arc.lock().await;
            if session.account.username == username {
                session.set_proxy(new_proxy.clone());
            }
        }
    }

    fn install_game_finder_callbacks(self: &Arc<Self>) {
        let weak = Arc::downgrade(self);
        let active_weak = Arc::downgrade(self);
        let mut guard = self.game_finder.lock().expect("game finder mutex poisoned");
        let Some(game_finder) = guard.as_mut() else {
            return;
        };

        game_finder.set_observation_starter(Arc::new(move |game_id, scenario_id| {
            if let Some(observer) = weak.upgrade() {
                observer.scheduler.queue_new_session(game_id, scenario_id);
            }
        }));

        game_finder.set_active_session_counter(Arc::new(move || {
            active_weak
                .upgrade()
                .map(|observer| {
                    observer
                        .observer_sessions
                        .lock()
                        .expect("observer sessions mutex poisoned")
                        .len()
                })
                .unwrap_or(0)
        }));

        tracing::info!("installed GameFinder callbacks");
    }

    async fn install_account_pool_callbacks(self: &Arc<Self>) {
        let weak = Arc::downgrade(self);
        let callback = Arc::new(move |username: String| {
            let Some(observer) = weak.upgrade() else {
                return;
            };
            tokio::spawn(async move {
                let new_proxy = {
                    let pool = observer.account_pool.lock().await;
                    pool.get_account_proxy(&username)
                };
                if let Some(proxy) = new_proxy {
                    observer.reset_sessions_for_account(&username, proxy).await;
                }
            });
        });

        let mut pool = self.account_pool.lock().await;
        pool.set_proxy_reset_callback(callback);
        tracing::info!("installed AccountPool proxy reset callback");
    }
}

